mlir/lib/QEC/Utils/QECLayer.cpp

```diff
@@ -0,0 +1,246 @@
+// Copyright 2025 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/ADT/STLExtras.h"
+
+#include "QEC/Utils/PauliStringWrapper.h"
+#include "QEC/Utils/QECLayer.h"
+#include "Quantum/IR/QuantumOps.h" // for quantum.extract op
+
+using namespace catalyst::qec;
+
+namespace catalyst {
+namespace qec {
+
+void QECLayer::insertToLayer(QECOpInterface op)
+{
+    ops.emplace_back(op);
+    updateResultAndOperand(op);
+
+    // Update the cached entry qubit set when inserting
+    auto entryQubits = getEntryQubitsFrom(op);
+    layerEntryQubits.insert(entryQubits.begin(), entryQubits.end());
+}
+
+void QECLayer::eraseOp(QECOpInterface op) { llvm::erase(ops, op); }
+
+void QECLayer::updateResultAndOperand(QECOpInterface op)
+{
+    // Ensure layer operand set contains canonical origins for any input qubits
+    llvm::SmallVector<Value> inQubits(op.getInQubits().begin(), op.getInQubits().end());
+    llvm::SmallVector<Value> outQubits(op.getOutQubits().begin(), op.getOutQubits().end());
```

Comment:
I think you could use ValueRange here to remove redundancy
```suggestion
    ValueRange inQubits = op.getInQubits();
    ValueRange outQubits = op.getOutQubits();
```